### DISCUSSION TITLE
How to generate mipmaps?

### DISCUSSION THREAD

**Discussion by gamma-delta** [malcolmriley/unused-renders#8] [Ideas and Requests]
Hello! I've been following these tutorials to make some sprites for a mod I'm working on and they're excellent.

I'm getting ready to export my first sprite into the game, and so I'd like to generate some mipmaps. Is there an automated way to do this with Blender or ImageMagick? It looks like the tutorials kind of skip over how to actually *export* the images.

**Comment by gamma-delta** [comment]
I ended up figuring this out:

```fish
magick -background none $IMG'[64x64]' $IMG'[32x32]' $IMG'[16x16]' $IMG'[8x8]' +append blank-data-card.png
```

This uses Fish shell; Bash probably has different escaping rules.

I'm not sure of a better way to avoid typing that all out :<

**Reply by malcolmriley** [reply]
Hello,

Your solution is how I do it as well - a shell script invokes ImageMagick to reduce the scale of a sprite and append it repeatedly. The contents of the script I used to generate the sprites in this repository is as follows:

```zsh
magick -background none ${1} -trim -gravity center -filter triangle -resize 64x64 -extent 64x64 \
	-gravity north \( +clone -resize 32x32 \) \( +clone -resize 16x16 \) \( +clone -resize 8x8 \) +append mipped/${1:t}
```

This is using ZSH/Bash syntax, so some adaptation may be required (I am unfamiliar with Fish).

You certainly noticed a few extra ImageMagick parameters in there:

- `-filter triangle` sets the resampling filter for reduction ([Documentation Here](https://imagemagick.org/script/command-line-options.php#filter))
- `-trim` is an operation that removes excess whitespace around the sprite (trimming the canvas so that the outermost non-transparent pixel is flush with the image boundary ([Documentation Here](https://imagemagick.org/script/command-line-options.php#trim))
- `+clone` copies the image on the internal operation stack
- `-gravity north` sets the anchor point for appending ([Documentation Here](https://imagemagick.org/script/command-line-options.php#gravity))
- The escaped parentheses are a "set-aside" operator

The "preprocessing" bits (for trimming and recentering) can be skipped if you're working from an image that already has those steps completed.

I haven't used Fish myself, but according to the documentation it has a similar feature to ZSH/Bash wherein the arguments passed to saved scripts are contained in an array called `$argv` (so accessing the first argument would be `$argv[1]` - I'm not certain but I believe `$argv[0]` is the "zeroth" argument - the invoked command itself).

It's worth noting that Factorio now automatically generates mipmaps (see [here](https://lua-api.factorio.com/latest/types/IconData.html)), so this extra step is not necessary unless you need/want more fine-grained control of the resampling step (Factorio uses Trilinear interpolation when generating mipmaps automatically([as mentioned in FFF-291](https://factorio.com/blog/post/fff-291)). Manual control of the mipmapping step is good for situations where one wants a highly detailed sprite for up-close (such as inventory view), but would prefer to supply a different, low-detail sprite for zoom-out where excess detail would look muddy or noisy (such as for items on belts).

I hope that answered your questions. If you have any further, please don't hesitate to ask! Good luck with your mod!

**Reply by gamma-delta** [reply]
Thanks for the quick reply! This repo's been invaluable; I'm working on some ice ore textures now using geometry nodes, and if I feel especially frisky I might write up an article on how I did it and submit it.

**Reply by malcolmriley** [reply]
Since you mentioned an ice ore - if you'd like to take a peek at how I've been doing ores and translucent items, I've isolated a few example objects from my project into an attached archive of two standalone Blender files. It's not as tidy as a file I'd include with a tutorial but I hope you can at least use it as inspiration to create something better. Consider this material to be released unencumbered as `CC-0`, public domain (material here meaning the attached files and content therein and not merely the Blender `Material`) .

The general explanation is as follows:
- **Ores:** Scatter some procedural meshes onto a plane using geometry nodes, exploiting animation keyframes to automatically generate ore variants - this dovetails nicely with command-line rendering as one can specify the exact frame to render, then render in a batch and composite as a single step with the right script. In this example, the ground texture overlay is imperfect; it currently lacks the necessary dither to allow the sprite to blend with the terrain (examine the vanilla files to see what I mean). A complete implementation would be to render the ground texture overlay as a cutout and then during the postprocessing step apply dither to only this overlay and then recomposite with the corresponding ore sprite. When I work out a nice workflow for this (and when I have some spare time) I'll add a tutorial.
- **Translucent Items:** Some combination of the `Glass BSDF` and `Principled BSDF`, using geometry nodes to scatter some other meshes for internal details/imperfections (like bubbles in glass, for instance). The most important insight here is to flip the normals for geometry intended to be internal detail of a translucent object so that the ray pathing is correct - it needs to deflect rays like a bubble of air inside an object rather than would another translucent object inside a translucent object. Examples of other phenomena like chromatic aberration are included (see the diamond scene and material). Chromatic aberration isn't correct-to-reality in this case of course but I think it looks nice!

[Translucent-Ore-Examples.zip](https://github.com/user-attachments/files/18142134/Translucent-Ore-Examples.zip)

**Reply by malcolmriley** [reply]
I've just realized that the above Blender file(s) contain broken links to an HDR I've been using ([it is this one](https://polyhaven.com/a/abandoned_workshop)).

This error will cause seemingly purple textures for the reflective materials. This can be corrected by supplying a link to a new HDR in the scene environment tab, or by using one of the built-in Sky texture generators.

I apologize for any inconvenience.
